compiler/src/middle_end/matchcomp.re

```diff
@@ -719,120 +368,200 @@ let rec rotate_if_needed = mtx =>
     switch (row) {
     | [] => failwith("Impossible: Empty pattern row in specialize_matrix")
     | [{pat_desc} as p, ...ptl] =>
       switch (pat_desc) {
-      | _ when pattern_always_matches(p) =>
-        let wildcards = List.init(arity, _ => {...p, pat_desc: TPatAny});
-        [wildcards @ ptl];
-      | TPatArray(args) when arity == List.length(args) => [args @ ptl]
+      | TPatAlias(p, alias, _) =>
+        specialized_rows([p, ...ptl], [(alias, cur), ...binds])
+      | TPatVar(id, _) =>
+        let wildcards = Parmatch.omegas(arity);
+        [(wildcards @ ptl, [(id, cur), ...binds])];
+      | TPatArray(args) when arity == List.length(args) => [
+          (args @ ptl, binds),
+        ]
       | TPatOr(p1, p2) =>
-        specialized_rows([p1, ...ptl]) @ specialized_rows([p2, ...ptl])
-      | TPatAlias(p, _, _) => specialized_rows([p, ...ptl])
+        specialized_rows([p1, ...ptl], binds)
+        @ specialized_rows([p2, ...ptl], binds)
+      | _ when pattern_always_matches(p) =>
+        let wildcards = Parmatch.omegas(arity);
+        [(wildcards @ ptl, binds)];
       | _ => [] /* Specialization for non-arrays generates no rows. */
       }
     };
 
   List.fold_right(
-    ((row, mb), rem) =>
-      List.map(patts => (patts, mb), specialized_rows(row)) @ rem,
+    ((row, binds, mb), rem) =>
+      List.map(
+        ((patts, binds)) => (patts, binds, mb),
+        specialized_rows(row, binds),
+      )
+      @ rem,
     mtx,
     [],
   );
 };
 
-/* [See paper for details]
+let rec specialize_constant_matrix = (const, cur, mtx) => {
+  let rec specialized_rows = (row, binds) =>
+    switch (row) {
+    | [] => failwith("Impossible: Empty pattern row in specialize_matrix")
+    | [{pat_desc} as p, ...ptl] =>
+      switch (pat_desc) {
+      | TPatAlias(p, alias, _) =>
+        specialized_rows([p, ...ptl], [(alias, cur), ...binds])
+      | TPatVar(id, _) =>
+        let wildcard = Parmatch.omega;
+        [([wildcard, ...ptl], [(id, cur), ...binds])];
+      | TPatConstant(c) when c == const =>
+        let wildcard = Parmatch.omega;
+        [([wildcard, ...ptl], binds)];
+      | TPatOr(p1, p2) =>
+        specialized_rows([p1, ...ptl], binds)
+        @ specialized_rows([p2, ...ptl], binds)
+      | _ when pattern_always_matches(p) => [(row, binds)]
+      | _ => [] /* Specialization for non-constants generates no rows. */
+      }
+    };
+
+  List.fold_right(
+    ((row, binds, mb), rem) =>
+      List.map(
+        ((patts, binds)) => (patts, binds, mb),
+        specialized_rows(row, binds),
+      )
+      @ rem,
+    mtx,
+    [],
+  );
+};
+
+/* [See paper for details; modified slightly]
 
       Row: ([p_1; p_2; ...], a_j)
 
       Pattern p_1 (for row j)       Row(s) of D(P)
       -----------------------       --------------
            c(q_1,...,q_n)               No row
-                 _                [([p_2; ...], a_j)]
+                 _                [([p_1; p_2; ...], a_j)]
```

Comment:
This confuses me. The default matrix is supposed to drop the first pattern (that's the whole idea, as far as I understand). Can you explain?

Reply:
I like to think of `default_matrix` as being the same as any of the other specialization matrixes, except this one specializes on "any" patterns, but otherwise it's no different—just needs to select the rows where the pattern in the first column always matches. Because it knows that the first column always matches, when it selects a row it _can_ remove that first pattern, but it doesn't have to. In our case, we want to see that pattern since it might contain a binding that we need to handle. Removing that first pattern can potentially make compiling the matrix faster since it never needs to look at that column again, but it makes no difference for the compiled matrix.

The key bit is that we'll always select a row if all of the patterns in that row always match.

With how we handle our value stack, it's much simpler to leave the pattern there since we won't need to manage removing that value from the stack (and dealing with finding that value later when we want to use it in one of the branches). So now, when we reach a leaf, we've got the full expanded row and all of the bindings to go with it! 🎉

Reply:
I can get behind your rationale, but maybe we should have a comment explaining any divergence from the paper (since the paper's pattern syntax is a bit more trimmed down, the formula they provide is different from what's now shown in this comment).

Reply:
Makes sense to me to explain the above in the comment (as @peblair requested).

Reply:
LGTM